File: packages/client/src/kernel.ts
// Daemon ↔ kernel messages (docs/protocol.md §4). The TypeScript side never
// speaks this protocol; the types exist so fixtures are checked in every language.

import type { CellResult, ChildInfo, HandleInfo, SkillInfo, WireError } from "./protocol.ts"

export type ToKernelMessage =
  | { type: "execute"; id: string; code: string; timeout_ms: number }
  | { type: "interrupt" }
  | { type: "host_response"; id: string; result: unknown }
  | { type: "host_response"; id: string; error: WireError }
  | { type: "shutdown" }

export interface HostRequests {
  "rlm.spawn": { params: { task: string; name: string; model?: string | null }; result: ChildInfo }
  "rlm.list_subagents": { params: Record<string, never>; result: { children: ChildInfo[] } }
  "rlm.delete_subagent": { params: { child_id: string } | { name: string }; result: { child: ChildInfo } }
  "agent_message.send": {
    params: { message: string; receiver_role: "parent" | "child"; receiver_name?: string | null }
    result: { event_id: string }
  }
  "notice.bash_finished": {
    params: { handle_id: string; pid: number; exit_code: number | null; command: string }
    result: { event_id: string }
  }
  "notice.withdraw": { params: { event_id: string }; result: { withdrawn: boolean } }
  "skills.list": { params: Record<string, never>; result: { skills: SkillInfo[] } }
}

export type HostRequestMethod = keyof HostRequests

export const HOST_REQUEST_METHODS: readonly HostRequestMethod[] = [
  "rlm.spawn",
  "rlm.list_subagents",
  "rlm.delete_subagent",
  "agent_message.send",
  "notice.bash_finished",
  "notice.withdraw",
  "skills.list",
]

export type LogLevel = "debug" | "info" | "warn" | "error"

export type FromKernelMessage =
  | { type: "ready"; pid: number; python: string; protocol: number }
  | { type: "execute_result"; id: string; result: CellResult }
  | { type: "host_request"; id: string; method: HostRequestMethod; params: unknown }
  | { type: "handles"; handles: HandleInfo[] }
  | { type: "log"; level: LogLevel; message: string }

File: crates/protocol/src/kernel.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{CellResult, RpcError};

/// Daemon → kernel protocol line.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ToKernel {
    Execute {
        id: String,
        code: String,
        timeout_ms: u64,
    },
    Interrupt,
    HostResponse {
        id: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        result: Option<Value>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        error: Option<RpcError>,
    },
    Shutdown,
}

/// Kernel → daemon protocol line.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum FromKernel {
    Ready {
        pid: u32,
        python: String,
        protocol: u32,
    },
    ExecuteResult {
        id: String,
        result: CellResult,
    },
    HostRequest {
        id: String,
        method: String,
        #[serde(default)]
        params: Value,
    },
    Handles {
        handles: Vec<HandleInfo>,
    },
    Log {
        level: String,
        message: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct HandleInfo {
    pub handle_id: String,
    pub pid: u32,
    pub command: String,
    pub running: bool,
}
